### PR TITLE
Thomson and Bremss opacities

### PR DESCRIPTION
**Pull request by brryan**
I add photon variants for Thomson scattering opacity and electron-proton bremsstrahlung opacity. 

Along the way I add a mean scattering class to photons analogous to the recently added mean scattering class for neutrinos #34

### CODE REVIEW

File: singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    const Real ni = ne;
+    return 4. * M_PI * pc::kb * temp / pc::h * prefac_ / std::sqrt(thetaE) *
+           ne * ni * gff_;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real NumberEmissivity(const Real rho, const Real temp,
+                        Real *lambda = nullptr) const {
+    // Infrared catastrophe
+    return std::numeric_limits<Real>::infinity();
```
Comment (brryan) [inline singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp:133]: \int d\Omega d\nu (1 / (h \nu)) exp^{-h \nu / (k T)}

Comment (Yurlungur) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp:133]: lol :+1:

File: singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+ private:
+  Real mass_ion_;
+  PlanckDistribution<pc> dist_;
+  static constexpr Real mmw_ =
+      (pc::mp + pc::me) /
+      (2. * pc::mp); // Neutral fully ionized electron-proton gas
+  static constexpr Real gff_ = 1.2;
+  static constexpr Real prefac_ = 8. * std::pow(pc::qe, 6) /
+                                  (3. * std::pow(pc::me * pc::c * pc::c, 2)) *
```
Comment (Yurlungur) [inline singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: didn't know `std::pow` was constexpr... nice.

File: singularity-opac/photons/s_opac_photons.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <singularity-opac/photons/gray_s_opacity_photons.hpp>
+#include <singularity-opac/photons/non_cgs_s_photons.hpp>
+#include <singularity-opac/photons/photon_s_variant.hpp>
+#include <singularity-opac/photons/thomson_s_opacity_photons.hpp>
+
+namespace singularity {
+namespace photons {
+
+// TODO(JMM): Include chemical potential
```
Comment (Yurlungur) [inline singularity-opac/photons/s_opac_photons.hpp]: probably not relevant for photons

**Review by Yurlungur — APPROVED** [review]
My reservations about the amount of boiler plate continue to grow, but otherwise looks good. 

I wonder if we could re-use some of the mean opacity stuff accross photons/neutrinos, because it's ALMOST identical, modulo Ye.

File: singularity-opac/photons/s_opac_photons.hpp (hunk 3 above)
Comment (brryan) [reply singularity-opac/photons/s_opac_photons.hpp]: Right thanks, fixed

File: singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp (hunk 2 above)
Comment (brryan) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: Actually this is a good point, it looks like `std::pow` being `constexpr` is an explicitly standards-breaking feature of gcc. I should go ahead and change this now before it causes problems later.

Comment (brryan) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: OK just removed the `static constexpr` decorators. In the future if we do want expressions like this to be `constexpr` we could add a dependency to [Sprout](https://github.com/bolero-MURAKAMI/Sprout)

**Comment by brryan** [comment]
Re: boilerplate, I agree, though I'm not immediately sure what the right thing to do is, and also I at least right now don't anticipate creating any more variants :)

File: singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp (hunk 2 above)
Comment (Yurlungur) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: these are also integer powers, so we could just write them out, or even implement our own constexpr pow. Something along the lines of the code here:
https://stackoverflow.com/questions/17719674/c11-fast-constexpr-integer-powers

Comment (Yurlungur) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: From stack overflow:
```C++
#include <cstdint>

constexpr int64_t ipow(int64_t base, int exp, int64_t result = 1) {
  return exp < 1 ? result : ipow(base*base, exp/2, (exp % 2) ? result*base : result);
}

int64_t foo(int64_t base, int exp) {
  return ipow(base, exp);
}

```

Comment (Yurlungur) [reply singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp]: Something like... (checked in godbolt):
```C++
constexpr Real mypow(Real x, std::size_t p) {
    if (p == 0) return 1;
    if (p == 1) return x;
    if ((p % 2) == 0) return mypow(x, p/2)*mypow(x, p/2);
    return mypow(x, p - 1)*x;
}
```

**Pull request merged**